pkg/tfvars/aws/aws.go

```diff
@@ -3,28 +3,43 @@ package aws
 
 import (
 	"encoding/json"
+	"fmt"
 
 	"github.com/pkg/errors"
 	"sigs.k8s.io/cluster-api-provider-aws/pkg/apis/awsproviderconfig/v1beta1"
 )
 
 type config struct {
-	EC2AMIOverride string            `json:"aws_ec2_ami_override,omitempty"`
-	ExtraTags      map[string]string `json:"aws_extra_tags,omitempty"`
-	EC2Type        string            `json:"aws_master_ec2_type,omitempty"`
-	IOPS           int64             `json:"aws_master_root_volume_iops"`
-	Size           int64             `json:"aws_master_root_volume_size,omitempty"`
-	Type           string            `json:"aws_master_root_volume_type,omitempty"`
-	Region         string            `json:"aws_region,omitempty"`
+	EC2AMIOverride    string            `json:"aws_ec2_ami_override,omitempty"`
+	ExtraTags         map[string]string `json:"aws_extra_tags,omitempty"`
+	AvailabilityZones []string          `json:"aws_master_availability_zones"`
+	EC2Type           string            `json:"aws_master_ec2_type,omitempty"`
+	IOPS              int64             `json:"aws_master_root_volume_iops"`
+	Size              int64             `json:"aws_master_root_volume_size,omitempty"`
+	Type              string            `json:"aws_master_root_volume_type,omitempty"`
+	Region            string            `json:"aws_region,omitempty"`
 }
 
 // TFVars generates AWS-specific Terraform variables launching the cluster.
-func TFVars(masterConfig *v1beta1.AWSMachineProviderConfig) ([]byte, error) {
+func TFVars(masterConfigs []*v1beta1.AWSMachineProviderConfig) ([]byte, error) {
+	masterConfig := masterConfigs[0]
+
 	tags := make(map[string]string, len(masterConfig.Tags))
 	for _, tag := range masterConfig.Tags {
 		tags[tag.Name] = tag.Value
 	}
 
+	availabilityZones := make([]string, len(masterConfigs))
+	azDups := make(map[string]bool, len(masterConfigs))
+	for i, c := range masterConfigs {
+		az := c.Placement.AvailabilityZone
+		availabilityZones[i] = az
+		if azDups[az] {
+			return nil, fmt.Errorf("cannot have multiple masters in the same availability zone (%q)", az)
```

Comment:
Hmm, I think we might have to allow users to create 3 masters in 2 availability zones...

Reply:
@eparis contends otherwise. See https://coreos.slack.com/archives/C68TNFWA2/p1548261027194200?thread_ts=1548258448.193100&cid=C68TNFWA2. I don't have a strong opinion. There is more work to be done to support multiple masters in an AZ, though.

Reply:
We either have to allow multiple masters in a single AZ, or we need to dynamically determine the correct number of masters to run because not all regions have at least three AZs:

![image](https://user-images.githubusercontent.com/368848/52889251-f54bf600-3133-11e9-91c5-57b6375ab57f.png)

Dynamically determining the correct number of masters is going to be very difficult. Not to mention the fact that an even number of masters (more specifically, etcd nodes) offers no improvements to availability. In us-west-1, for example, we would have to go with a single master node. There's just no way we are going to get away with doing that.

We have to allow multiple masters in a single AZ in some scenarios.

Reply:
But I would be happy to be dissuaded of the idea that there is more work involved to support multiple masters. My thought was that either (1) the load balancer would need to be adjusted to support routing to multiple masters in the same subnet or (2) there would need to be 1-to-1 relationship between subnets and masters in an availability zone.

Reply:
We want three masters in regions with less than three zones.  Even if the setup is not robust against zone outages, we want three masters to handle rolling master upgrades (at some point ;).

Reply:
I don't think we'd need load balancer work.  We already support the bootstrap machine and a master in the same subnet.